src/evict/evict_lru.c

```diff
@@ -1539,44 +1541,44 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 		bytes_per_slot = 1 + cache_inuse / total_slots;
 		target_pages_clean = (uint32_t)(
 		    (btree_inuse + bytes_per_slot / 2) / bytes_per_slot);
-	} else
-		target_pages_clean = 0;
+	}
 
 	if (F_ISSET(cache, WT_CACHE_EVICT_DIRTY)) {
 		btree_inuse = __wt_btree_dirty_leaf_inuse(session);
 		cache_inuse = __wt_cache_dirty_leaf_inuse(cache);
 		bytes_per_slot = 1 + cache_inuse / total_slots;
 		target_pages_dirty = (uint32_t)(
 		    (btree_inuse + bytes_per_slot / 2) / bytes_per_slot);
-	} else
-		target_pages_dirty = 0;
+	}
 
 	target_pages = WT_MAX(target_pages_clean, target_pages_dirty);
 
-	if (target_pages == 0) {
-		/*
-		 * Randomly walk trees with a tiny fraction of the cache in
-		 * case there are so many trees that none of them use enough of
-		 * the cache to be allocated slots.  Walk small trees 1% of the
-		 * time.
-		 */
-		if (__wt_random(&session->rnd) > UINT32_MAX / 100)
-			return (0);
+	/*
+	 * Randomly walk trees with a small fraction of the cache in case there
+	 * are so many trees that none of them use enough of the cache to be
+	 * allocated slots.  Make the chance of walking a tree equal to the
+	 * chance that a random byte in cache belongs to the tree.
+	 */
+	if (target_pages == 0 && (cache_inuse == 0 ||
+	    __wt_random64(&session->rnd) % cache_inuse > btree_inuse))
```

Comment:
If both dirty and clean eviction are enabled ``cache_inuse`` and ``btree_inuse`` reflect the dirty proportion of data.

I'm struggling a bit to see how this calculation actually works. I feel like the number of trees should play into the left side of the calculation as well. If there are 1000 trees in a 100GB cache. 800 type A trees use 1MB (round to 0), 100 type B trees use 100MB (10GB), 50 type C trees use 500MB (25GB) and 50 type D trees use 1.3GB each.

| Tree type | Frequency of selection (one selection per X passes) | Number of selections per traverse of handles |
|---|---|---|
| Type A | 100,000 | 0.008 |
| Type B | 1000 | 0.01 |
| Type C | 200 | 0.25 |
| Type D | 76 | 0.64 |

Frequency of selection = Cache in use / (# trees of type * tree size)
Number of selections = Frequency of selection / count of similar handles